torch/optim/sparse_adam.py

```diff
@@ -15,25 +15,21 @@ def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8, maximize: bool
         if not 0.0 <= betas[1] < 1.0:
             raise ValueError(f"Invalid beta parameter at index 1: {betas[1]}")
 
-        params = list(params)
+        defaults = dict(lr=lr, betas=betas, eps=eps, maximize=maximize)
+        super().__init__(params, defaults)
 
         sparse_params = []
-        for index, param in enumerate(params):
-            if isinstance(param, dict):
-                # given param group, convert given params to a list first before iterating
-                param['params'] = list(param.get("params", []))
-                for d_index, d_param in enumerate(param['params']):
-                    if d_param.is_sparse:
-                        sparse_params.append([index, d_index])
-            elif param.is_sparse:
-                sparse_params.append(index)
```

Comment:
is this last branch handeld somewhere?

Reply:
This shouldn't be, since the param groups we build have a certain structure. That's why there is the added assert in line 23.